### PR TITLE
Naf obj #127

### PR DESCRIPTION
**Pull request by namirfa**
Fixes Issue #127 

Adds the ability to specify "savedResolution" options in the source configuration. This field contains an option called "longEdge", which is used to resize the saved image. The longEdge must be a non-negative integer that is smaller than the long edge of the original image. The saved image will be resized according to the longEdge value, and maintain the same dimension ratio as the original image.

The tests utilize the vantiq.download feature to check the dimensions for both locally saved and VANTIQ saved images.

This feature has not been added to the query parameters yet. It will be added as part of the work required for Issue #126 .

### CODE REVIEW

**Comment by namirfa** [comment]
I think I can add more of the ImageUtil changes into one method, since the code checking the width vs. length etc. is redundant.

I'll work on that now.

File: objectRecognitionSource/README.md [hunk 1]
```diff
@@ -222,6 +228,13 @@ is set to be either "local" or "both".
 captured). If not specified, the value will default to 1 which saves every captured image.
 *   labelImage: Optional. If set to "true", images will be saved with bounding boxes and labels. If set to "false", or if not 
 set at all, the images will be saved with no bounding boxes or labels.
+*   savedResolution: Optional. A map containing options for adjusting the resolution of the saved images.
+    * *Options for savedResolution:*
+    1. longEdge: Optional. Used to resize the saved images. Must be a non-negative integer that is smaller than the long edge
```
Comment (fhcarter) [inline objectRecognitionSource/README.md]: I think I'd change _Used to resize saved _images._ to _This sets the maximum long edge dimension for saved images._. By using this phraseology, we establish our _contract_ which then means that we don't have to worry about what happens if we save things that are smaller (because they were smaller), etc.

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java [hunk 2]
```diff
@@ -63,7 +64,33 @@ public void saveImage(final BufferedImage image, final String target) {
...
             try {
                 IOUtil.createDirIfNotExists(new File(outputDir));
-                ImageIO.write(image,"jpg", new File(outputDir + File.separator + target));
+                if (longEdge == 0) {
+                    ImageIO.write(image, "jpg", new File(outputDir + File.separator + target));
+                } else {
+                    int width = image.getWidth();
+                    int height = image.getHeight();
+                    if (longEdge > Math.max(width, height)) {
+                        LOGGER.error("The longEdge value is too large, resizing cannot enlarge the original image. "
```
Comment (fhcarter) [inline objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java]: I think we should make this a trace() level note.  We've fulfilled our contract (see comment above on the readme) so there's nothing amiss here.

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java [hunk 3]
```diff
@@ -63,7 +64,33 @@ public void saveImage(final BufferedImage image, final String target) {
...
+                                + "Original image resolution will not be changed");
+                        ImageIO.write(image, "jpg", new File(outputDir + File.separator + target));
+                        longEdge = 0;
```
Comment (fhcarter) [inline objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java]: This unsets longEdge for everything. Couldn't some source send back various images with different dimensions?  Seems like we just want to skip the resize, not change the expected behavior.

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java [hunk 4]
```diff
@@ -89,7 +116,33 @@ public void saveImage(final BufferedImage image, final String target) {
...
+                                double ratio = (double) longEdge/height;
+                                int newWidth = (int) (width * ratio);
+                                BufferedImage resizedImage = resizeImage(newWidth, longEdge, image);
+                                ImageIO.write(resizedImage, "jpg", imgFile);
+                            } else {
+                                BufferedImage resizedImage = resizeImage(longEdge, longEdge, image);
+                                ImageIO.write(resizedImage, "jpg", imgFile);
+                            }
+                        }
+                    }
```
Comment (fhcarter) [inline objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java]: Seems like we should be able to refactor all this stuff in a method to call.  I think you're working on this.

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java [hunk 5]
```diff
@@ -89,7 +116,33 @@ public void saveImage(final BufferedImage image, final String target) {
...
+                    if (longEdge == 0) {
+                        ImageIO.write(image, "jpg", imgFile);
+                    } else {
+                        int width = image.getWidth();
+                        int height = image.getHeight();
+                        if (longEdge > Math.max(width, height)) {
+                            LOGGER.error("The longEdge value is too large, resizing cannot enlarge the original image. "
+                                    + "Original image resolution will not be changed");
+                            ImageIO.write(image, "jpg", imgFile);
+                            longEdge = 0;
```
Comment (fhcarter) [inline objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java]: again, don't kill it forever

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java [hunk 6]
```diff
@@ -115,6 +168,21 @@ public void saveImage(final BufferedImage image, final String target) {
...
     }
+    
+    /**
+     * A function used to resize the original captured image, if requested.
+     * @param width     The resized width of the image to be saved.
+     * @param height    The resized height of the image to be saved.
+     * @param image     The original image to be resized.
+     * @return          The resized image.
+     */
+    public BufferedImage resizeImage(int width, int height, BufferedImage image) {
```
Comment (fhcarter) [inline objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java]: maybe we could have a variant that includes the longEdge that does all that work.

Comment (namirfa) [reply objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java]: Yep that's exactly what I'm doing now.

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java (hunk 4 above)
Comment (namirfa) [reply objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java]: Yep, I'm working on this.

File: objectRecognitionSource/README.md [hunk 7]
```diff
@@ -222,6 +228,13 @@ is set to be either "local" or "both".
 captured). If not specified, the value will default to 1 which saves every captured image.
 *   labelImage: Optional. If set to "true", images will be saved with bounding boxes and labels. If set to "false", or if not 
 set at all, the images will be saved with no bounding boxes or labels.
+*   savedResolution: Optional. A map containing options for adjusting the resolution of the saved images.
+    * *Options for savedResolution:*
+    1. longEdge: Optional. Used to resize the saved images. Must be a non-negative integer that is smaller than the long edge 
+    of the image to be saved. This value will become the new long edge dimension, and the short edge will be scaled down to 
+    maintain the same dimension ratio as the original image.
```
[the remark this answers is not in the document]
Comment (namirfa) [reply objectRecognitionSource/README.md]: I'll change that to aspect ratio. And we do have a test for the value being larger than the image. This is located in `TestYoloProcessor.testInvalidResizingSavedImage()`.

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java (hunk 3 above)
Comment (namirfa) [reply objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java]: Just talked about this, forget the comment above :).
